test/unit/pipeline/FetchUnitTest.cc

```diff
@@ -437,5 +453,238 @@ TEST_F(PipelineFetchUnitTest, idleLoopBufferDueToNotTakenBoundary) {
   EXPECT_EQ(output.getTailSlots()[0], mOp2);
 }
 
+// Tests that a compressed instruction held at the end of the fetch buffer is
+// allowed to be predecoded in the same cycle as being fetched
+TEST_F(PipelineFetchUnitTest, compressedInstructionAtEndOfBuffer) {
+  ON_CALL(isa, getMaxInstructionSize()).WillByDefault(Return(insnMaxSizeBytes));
+  ON_CALL(isa, getMinInstructionSize()).WillByDefault(Return(insnMinSizeBytes));
+  ON_CALL(memory, getCompletedReads()).WillByDefault(Return(completedReads));
+
+  // Buffer will contain valid compressed instruction so predecode returns
+  // 2 bytes read
+  MacroOp mOp = {uopPtr};
+  ON_CALL(isa, predecode(_, 2, 0xE, _))
+      .WillByDefault(DoAll(SetArgReferee<3>(mOp), Return(2)));
+
+  // Fetch the data, only 2 bytes will be copied to fetch buffer. Should allow
+  // continuation to predecode
+  EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+  EXPECT_CALL(memory, getCompletedReads()).Times(1);
+  EXPECT_CALL(isa, getMinInstructionSize()).Times(1);
+  EXPECT_CALL(isa, predecode(_, _, _, _)).Times(1);
+
+  // Set PC to 14, buffered bytes = 0
+  fetchUnit.updatePC(14);
+  // Tick. Fetch data and predecode
+  fetchUnit.tick();
+
+  // Buffer should now be empty as all bytes predecoded
+  EXPECT_EQ(fetchUnit.bufferedBytes_, 0);
+
+  // Expect a block starting at address 16 to be requested when we fetch again
+  EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+  EXPECT_CALL(memory, requestRead(Field(&MemoryAccessTarget::address, 16), _))
+      .Times(1);
+  fetchUnit.requestFromPC();
+
+  // Tick again, expecting that decoding will now resume
+  MemoryReadResult nextBlockValue = {{16, blockSize}, 0, 1};
+  span<MemoryReadResult> nextBlock = {&nextBlockValue, 1};
+  ON_CALL(memory, getCompletedReads()).WillByDefault(Return(nextBlock));
+  ON_CALL(isa, predecode(_, _, _, _))
+      .WillByDefault(DoAll(SetArgReferee<3>(mOp), Return(4)));
+
+  EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+  // Completed reads called again as more data is requested
+  EXPECT_CALL(memory, getCompletedReads()).Times(1);
+  EXPECT_CALL(isa, getMinInstructionSize()).Times(1);
+  // Output of width 1 so only 1 call to predecode
+  EXPECT_CALL(isa, predecode(_, _, _, _)).Times(1);
+
+  fetchUnit.tick();
+
+  // Initially 0 bytes, 16 bytes added, 4 bytes predecoded leaving 12 bytes left
+  EXPECT_EQ(fetchUnit.bufferedBytes_, 12);
```

Comment:
We should complement these types of checks with ensuring the output buffer slot contains the expected macro-op

Reply:
+1